File: daily_problems/2026/03/0304/personal_submission/cf103316j_yawn_sean.py
# Submission link: https://codeforces.com/gym/103316/submission/365217742
def main(): 
    n = II()
    nums = LII()
    
    path = [[] for _ in range(n)]
    
    for _ in range(n - 1):
        u, v = GMI()
        path[u].append(v)
        path[v].append(u)
    
    st_range = sorted(range(n), key=lambda x: nums[x])
    
    ans = 0
    mod = 10 ** 9 + 7
    
    vis = [0] * n
    uf = UnionFind(n)
    
    for u in st_range:
        for v in path[u]:
            if vis[v]:
                ans = (ans + nums[u] * uf.getSize(u) % mod * uf.getSize(v) % mod) % mod
                uf.merge(u, v)
        vis[u] = 1
    
    for i in range(n): vis[i] = 0
    uf.init()
    
    for u in reversed(st_range):
        for v in path[u]:
            if vis[v]:
                ans = (ans + mod - nums[u] * uf.getSize(u) % mod * uf.getSize(v) % mod) % mod
                uf.merge(u, v)
        vis[u] = 1
    
    print(ans * pow(n * (n - 1) // 2 % mod, -1, mod) % mod)

File: daily_problems/2026/03/0304/personal_submission/cf103316j_yawn_sean.cpp
#include <bits/stdc++.h>
// #pragma GCC optimize("O3,Ofast,unroll-loops")
// #include "atcoder/all"
#include "atcoder/dsu"

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

// Let's unordered_set a bit and flow!
// I came, I divided, I conquered!
// using namespace __gnu_cxx;
// using namespace __gnu_pbds;

long long quickPow(long long base, long long power, long long mod) {
	if (power == 0) return 1 % mod;
	long long cur = quickPow(base, power / 2, mod);
	return power & 1 ? base * cur % mod * cur % mod : cur * cur % mod; 
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n;
	cin >> n;

	vector<int> nums(n);
	for (auto &v: nums) cin >> v;

	vector<vector<int>> path(n);
	for (int i = 0; i < n - 1; i ++) {
		int u, v;
		cin >> u >> v;
		u --, v --;
		path[u].emplace_back(v);
		path[v].emplace_back(u);
	}

	vector<int> order(n);
	iota(order.begin(), order.end(), 0);
	sort(order.begin(), order.end(), [&] (int i, int j) {return nums[i] < nums[j];});

	int ans = 0, mod = 1e9 + 7;

	atcoder::dsu uf(n);
	vector<int> vis(n, 0);

	for (auto &u: order) {
		for (auto &v: path[u]) {
			if (vis[v]) {
				ans += 1ll * nums[u] * uf.size(u) % mod * uf.size(v) % mod;
				ans %= mod;
				uf.merge(u, v);
			}
		}
		vis[u] = 1;
	}

	reverse(order.begin(), order.end());

	uf.init();
	fill(vis.begin(), vis.end(), 0);

	for (auto &u: order) {
		for (auto &v: path[u]) {
			if (vis[v]) {
				ans += mod - 1ll * nums[u] * uf.size(u) % mod * uf.size(v) % mod;
				ans %= mod;
				uf.merge(u, v);
			}
		}
		vis[u] = 1;
	}

	cout << 1ll * ans * quickPow(1ll * n * (n - 1) / 2 % mod, mod - 2, mod) % mod;

	return 0;
}